### PR TITLE
Adds Cashtoken support

### LINKED ISSUE #121: Cash Tokens support CHIP-2022-02-CashTokens
Would like to request support for Cash Tokens CHIP-2022-02-CashTokens

https://github.com/bitjson/cashtokens

### PR DESCRIPTION
**Pull request by yashasvi-ranawat**
Fixes #121, builds on #119.

Adds Cashtoken support to Bitcash.

- [x] Unspent are aware of cashtoken
- [x] Cashtoken handling class
- [x] Sanitized outputs are aware of cashtokens
- [x] wallets are aware of cashtokens
- [x] wallets can generate cashtoken tx
- [x] wallets can generate cashtokens
- [x] cashaddress signals cashtoken support
- [x] wallets signal cashtoken support 
- [x] add test vectors for cashtoken prefix and token aware cashaddress
- [x] add cashtoken documentation

### CODE REVIEW

**Comment by yashasvi-ranawat** [comment]
@merc1er I realised the CashToken spec has test vectors too, added them as well. Its ready for review.
The final thing left is to update the CASHTOKEN_ACTIVATION_BLOCKHEIGHT to the block after 1684152000 MTP.

**Comment by yashasvi-ranawat** [comment]
This Cashtoken pull request is designed with the assumption that the utxo query from the BitcoinDotComAPI will fetch scriptcodes with attached token prefix.
However, while adding refactoring PATFO check to the BitcoinDotComAPI, I realised the API sends only the locking script for all utxos (which makes sense). This implies that any token prefix attached to individual utxo might be missed.
Therefore, this pull request is dependant on https://devsuite.actorforth.org/ on adding Cashtoken support.

**Comment by merc1er** [comment]
@yashasvi-ranawat: Does the CashToken sections shows for you in the index page of the docs? I tried building locally (`cd docs; make html`) but I am not seeing anything about tokens.

**Comment by yashasvi-ranawat** [comment]
> @yashasvi-ranawat: Does the CashToken sections shows for you in the index page of the docs? I tried building locally (`cd docs; make html`) but I am not seeing anything about tokens.

Yes. I checked the changelog on this pull request, I do see the doc changes that should add it to the index page.

**Review by merc1er — CHANGES_REQUESTED** [review]
Outstanding work!

Thanks a lot for both fixing the UTXO issue and implementing the Cash Token feature at the same time!

As discussed, I left a comment on the address format. I also left a couple very minor comments.

File: docs/source/guide/cashtokens.rst [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+non-cashtoken-signalling addresses:
+
+.. code-block:: python
+
+   >>> key.address
+   'bitcoincash:zrweeythv25ltpdypewr54prs6zd3nr5rcjhrnhy2v'
+
+However, some wallets might not support the new CashAddr format yet. For such cases,
+function :func:`~bitcash.format.cashtokenaddress_to_address` is provided to convert to 
+the version which do not signal cashaddress.
```
[the remark this answers is not in the document]
Comment (yashasvi-ranawat) [reply docs/source/guide/cashtokens.rst]: Added key.cashtoken_address

File: bitcash/transaction.py [hunk 2]
```diff
@@ -26,31 +32,21 @@
...
 
 
 class TxIn:
-    __slots__ = ("script", "script_len", "txid", "txindex", "amount")
+    __slots__ = ("script", "script_len", "txid", "txindex", "amount", "token_prefix")
 
-    def __init__(self, script, script_len, txid, txindex, amount):
+    def __init__(self, script, script_len, txid, txindex, amount, token_prefix=None):
+        if token_prefix is None:
+            token_prefix = b""
```
[the remark this answers is not in the document]
Comment (yashasvi-ranawat) [reply bitcash/transaction.py]: Wrongly thought binary is mutable in python :/

File: bitcash/network/APIs/BitcoinDotComAPI.py [hunk 3]
```diff
@@ -53,13 +56,15 @@ def make_endpoint_url(self, path):
         return self.network_endpoint + self.PATHS[path]
 
     def get_balance(self, address, *args, **kwargs):
+        address = cashtokenaddress_to_address(address)
```
[the remark this answers is not in the document]
Comment (yashasvi-ranawat) [reply bitcash/network/APIs/BitcoinDotComAPI.py:59]: Changed the CashAddr format, however still left the function.

**Review by merc1er — APPROVED** [review]
I'm very impressed with the quality of your work. Everything that I tested worked flawlessly.

Thank you for this major contribution!

Could you please confirm that there no longer is anything not backwards compatible in this PR?

**Comment by merc1er** [comment]
Also if you don't mind I'll add you to the `AUTHORS.md` file as a contributor.

**Comment by merc1er** [comment]
For some reason tests won't pass on Python 3.7. That's not much of an issue (Python 3.7 was released over 5 years ago), but could you confirm we no longer support 3.7 with the newest changes?

If so I'll remove it in favor of 3.11, and release version `1.0.0` of BitCash.

**Comment by yashasvi-ranawat** [comment]
> Could you please confirm that there no longer is anything not backwards compatible in this PR?

I tested with pre-merge tests, wallet and cashaddress are backwards compatible. The sanitised output format has been changed, so no backwards compatibility in the transactions.py. However, those I think are not directly used by any user, but via wallet.py. The old output (the one user makes, not the sanitised output) format still works.

> For some reason tests won't pass on Python 3.7. That's not much of an issue (Python 3.7 was released over 5 years ago), but could you confirm we no longer support 3.7 with the newest changes?

Python 3.7 is already at its [end-of-life](https://devguide.python.org/versions/)

**Comment by bitjson** [comment]
@yashasvi-ranawat fantastic work, thanks for all you're doing on this library! Do you have a Twitter account I can follow/highlight?

**Comment by yashasvi-ranawat** [comment]
@bitjson Thanks a lot. My twitter handle: https://twitter.com/yashasviranawat

**Pull request merged**